Flaky test: test_check_table/test.py::test_check_all_tables

_Important: This issue was automatically generated and is used by CI for matching failures. DO NOT modify the body content. DO NOT remove labels._

Test name: test_check_table/test.py::test_check_all_tables
Failure reason: was created concurrently with remote path
CI report: [Integration tests (amd_asan_ubsan, db disk, old analyzer, 5/6)](https://s3.amazonaws.com/clickhouse-test-reports/json.html?PR=110449&sha=733564ba5596c5057a9b37d05b1ee3c1fc17dd30&name_0=PR&name_1=Integration%20tests%20%28amd_asan_ubsan%2C%20db%20disk%2C%20old%20analyzer%2C%205%2F6%29)

Failing test history: [cidb](https://play.clickhouse.com#V0lUSAogICAgOTAgQVMgaW50ZXJ2YWxfZGF5cwpTRUxFQ1QKICAgIHRvU3RhcnRPZkRheShjaGVja19zdGFydF90aW1lKSBBUyBkYXksCiAgICBjb3VudCgpIEFTIGZhaWx1cmVzLAogICAgZ3JvdXBVbmlxQXJyYXkocHVsbF9yZXF1ZXN0X251bWJlcikgQVMgcHJzLAogICAgYW55KHJlcG9ydF91cmwpIEFTIHJlcG9ydF91cmwKRlJPTSBjaGVja3MKV0hFUkUgKG5vdygpIC0gdG9JbnRlcnZhbERheShpbnRlcnZhbF9kYXlzKSkgPD0gY2hlY2tfc3RhcnRfdGltZQogICAgQU5EIHRlc3RfbmFtZSA9ICd0ZXN0X2NoZWNrX3RhYmxlL3Rlc3QucHk6OnRlc3RfY2hlY2tfYWxsX3RhYmxlcycKICAgIC0tIEFORCBjaGVja19uYW1lID0gJ0ludGVncmF0aW9uIHRlc3RzIChhbWRfYXNhbl91YnNhbiwgZGIgZGlzaywgb2xkIGFuYWx5emVyLCA1LzYpJwogICAgQU5EIHRlc3Rfc3RhdHVzIElOICgnRkFJTCcsICdFUlJPUicpCiAgICBBTkQgKHB1bGxfcmVxdWVzdF9udW1iZXIgPSAwIE9SIGJhc2VfcmVmIElOICgnbWFzdGVyJykpCiAgICBBTkQgdGVzdF9jb250ZXh0X3JhdyBMSUtFICcld2FzIGNyZWF0ZWQgY29uY3VycmVudGx5IHdpdGggcmVtb3RlIHBhdGglJwpHUk9VUCBCWSBkYXkKT1JERVIgQlkgZGF5IERFU0MK)

Test output:
```
File: test_check_table/test.py:90 - in test_check_all_tables
    future.result()
File: /usr/lib/python3.13/concurrent/futures/_base.py:453 - in result
    return self.__get_result()
           ^^^^^^^^^^^^^^^^^^^
File: /usr/lib/python3.13/concurrent/futures/_base.py:402 - in __get_result
    raise self._exception
File: /usr/lib/python3.13/concurrent/futures/thread.py:59 - in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File: test_check_table/test.py:87 - in <lambda>
    futures.append(executor.submit(lambda: node1.query("DROP DATABASE db2")))
                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File: helpers/cluster.py:5045 - in query
    return self.client.query(
File: helpers/client.py:40 - in wrap
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
File: helpers/client.py:80 - in query
    ).get_answer()
      ^^^^^^^^^^^^
File: helpers/client.py:269 - in get_answer
    raise QueryRuntimeException(
E   helpers.client.QueryRuntimeException: Client failed! Return code: 84, stderr: Received exception from server (version 26.7.1):
E   Code: 84. DB::Exception: Received from 172.16.3.9:9000. DB::Exception: Directory 'metadata_dropped' was created concurrently with remote path 'mbziyskezfsgjguanirvyyyduroltyqo': While committing metadata operation #0. Stack trace:
E   
E   0. contrib/llvm-project/libcxx/include/__exception/exception.h:115:14: Poco::Exception::Exception(String&&, int) @ 0x0000000060ee3fc0
E   1. src/Common/Exception.cpp:139:7: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000002ed69b48
E   2. src/Common/Exception.h:173:100: DB::Exception::Exception(String&&, int, String, bool) @ 0x000000001def5455
E   3. src/Common/Exception.h:57:54: DB::Exception::Exception(PreformattedMessage&&, int) @ 0x000000001def3dbe
E   4. src/Common/Exception.h:191:77: DB::Exception::Exception<String, String const&>(int, FormatStringHelperImpl<std::type_identity<String>::type, std::type_identity<String const&>::type>, String&&, String const&) @ 0x000000001f1230fa
E   5. src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/Transactions/Preconditions.cpp:36:15: DB::Preconditions::runChecks(std::shared_ptr<DB::FsSnapshot const> const&) @ 0x000000003f3d03f8
E   6. src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/MetadataStorageFromPlainRewritableObjectStorageOperations.cpp:58:20: DB::MetadataStorageFromPlainObjectStorageValidatePreconditionsOperation::execute() @ 0x000000003f3d38d8
E   7. src/Disks/DiskObjectStorage/MetadataStorages/MetadataOperationsHolder.cpp:72:28: DB::MetadataOperationsHolder::commit() @ 0x000000003f34a62d
E   8. src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/MetadataStorageFromPlainRewritableObjectStorage.cpp:412:20: DB::MetadataStorageFromPlainRewritableObjectStorageTransaction::commit(std::variant<std::monostate, DB::MetaInKeeperCommitOptions<zkutil::ZooKeeper>, DB::MetaInKeeperCommitOptions<DB::ZooKeeperWithFaultInjection>> const&) @ 0x000000003f38845c
E   9. src/Disks/DiskObjectStorage/DiskObjectStorageTransaction.cpp:566:31: DB::DiskObjectStorageTransaction::commit() @ 0x000000003f25318b
E   10. src/Disks/DiskObjectStorage/DiskObjectStorage.cpp:439:22: DB::DiskObjectStorage::createDirectories(String const&) @ 0x000000003f1ed124
E   11. src/Databases/DatabaseAtomic.cpp:218:18: DB::DatabaseAtomic::dropTableImpl(std::shared_ptr<DB::Context const>, String const&, bool) @ 0x000000003ee609ce
E   12. src/Databases/DatabaseAtomic.cpp:204:5: DB::DatabaseAtomic::dropTable(std::shared_ptr<DB::Context const>, String const&, bool) @ 0x000000003ee5ff21
E   13. src/Interpreters/InterpreterDropQuery.cpp:358:23: DB::InterpreterDropQuery::executeToTableImpl(std::shared_ptr<DB::Context const> const&, DB::ASTDropQuery&, std::shared_ptr<DB::IDatabase>&, StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>&) @ 0x0000000041bd97e7
E   14. src/Interpreters/InterpreterDropQuery.cpp:668:17: DB::InterpreterDropQuery::executeToDatabaseImpl(DB::ASTDropQuery const&, std::shared_ptr<DB::IDatabase>&, std::vector<StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>, std::allocator<StrongTypedef<wide::integer<128ul, unsigned int>, DB::UUIDTag>>>&) @ 0x0000000041bdfa34
E   15. src/Interpreters/InterpreterDropQuery.cpp:415:15: DB::InterpreterDropQuery::executeToDatabase(DB::ASTDropQuery const&) @ 0x0000000041bd5267
E   16. src/Interpreters/InterpreterDropQuery.cpp:120:16: DB::InterpreterDropQuery::executeSingleDropQuery(boost::intrusive_ptr<DB::IAST> const&) @ 0x0000000041bd2e44
E   17. src/Interpreters/InterpreterDropQuery.cpp:93:15: DB::InterpreterDropQuery::execute() @ 0x0000000041bd24b6
E   18. src/Interpreters/executeQuery.cpp:1895:40: DB::executeQueryImpl(char const*, char const*, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum, std::unique_ptr<DB::ReadBuffer, std::default_delete<DB::ReadBuffer>>&, boost::intrusive_ptr<DB::IAST>&, std::shared_ptr<DB::ImplicitTransactionControlExecutor>, std::function<void ()>, DB::QueryResultDetails&) @ 0x00000000425be543
E   19. src/Interpreters/executeQuery.cpp:2371:11: DB::executeQuery(std::basic_string_view<char, std::char_traits<char>>, std::shared_ptr<DB::Context>, DB::QueryFlags, DB::QueryProcessingStage::Enum) @ 0x00000000425b341e
E   20. src/Server/TCPHandler.cpp:892:68: DB::TCPHandler::runImpl() @ 0x000000004d461dc5
E   21. src/Server/TCPHandler.cpp:3314:9: DB::TCPHandler::run() @ 0x000000004d4ae02d
E   22. base/poco/Net/src/TCPServerConnection.cpp:40:3: Poco::Net::TCPServerConnection::start() @ 0x00000000610c27af
E   23. base/poco/Net/src/TCPServerDispatcher.cpp:115:42: Poco::Net::TCPServerDispatcher::run() @ 0x00000000610c365b
~~~~~ truncated 7 lines ~~~~~
```


<!-- ch-version-info:start -->
### Version info
- Resolved by: #111353
- Backported to: `26.7.2.1`
<!-- ch-version-info:end -->

--- a/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/Metadata/FsSnapshot.cpp
+++ b/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/Metadata/FsSnapshot.cpp
@@ -302,15 +302,15 @@ std::vector<std::string> FsSnapshot::listDirectory(const std::string & path) con
     return result;
 }
 
-std::pair<bool, std::optional<DirectoryRemoteInfo>> FsSnapshot::existsDirectory(const std::string & path) const
+bool FsSnapshot::existsDirectory(const std::string & path) const
 {
     UniqueLock lock(mutex);
     const auto node = walk(root, normalizePath(path));
 
     if (!node)
-        return {false, std::nullopt};
+        return false;
 
-    return {true, node->info};
+    return true;
 }
 
 std::unordered_map<std::string, std::optional<DirectoryRemoteInfo>> FsSnapshot::getSubtreeRemoteInfo(const std::string & path) const

--- a/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/Metadata/FsSnapshot.h
+++ b/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/Metadata/FsSnapshot.h
@@ -57,7 +57,7 @@ public:
     /// Directory Read Methods
 
     std::vector<std::string> listDirectory(const std::string & path) const;
-    std::pair<bool, std::optional<DirectoryRemoteInfo>> existsDirectory(const std::string & path) const;
+    bool existsDirectory(const std::string & path) const;
     std::unordered_map<std::string, std::optional<DirectoryRemoteInfo>> getSubtreeRemoteInfo(const std::string & path) const;
     std::optional<DirectoryRemoteInfo> getDirectoryRemoteInfo(const std::string & path) const;
 

--- a/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/MetadataStorageFromPlainRewritableObjectStorage.cpp
+++ b/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/MetadataStorageFromPlainRewritableObjectStorage.cpp
@@ -472,7 +472,6 @@ void MetadataStorageFromPlainRewritableObjectStorageTransaction::createDirectory
         metadata_storage.object_storage,
         metadata_storage.layout,
         metadata_storage.metrics));
-
 }
 
 void MetadataStorageFromPlainRewritableObjectStorageTransaction::moveDirectory(const std::string & path_from, const std::string & path_to)
@@ -486,12 +485,11 @@ void MetadataStorageFromPlainRewritableObjectStorageTransaction::moveDirectory(c
         metadata_storage.object_storage,
         metadata_storage.layout,
         metadata_storage.metrics));
-
 }
 
 void MetadataStorageFromPlainRewritableObjectStorageTransaction::unlinkFile(const std::string & path, bool if_exists, bool /*should_remove_objects*/)
 {
-    uncommitted_state.lookupDirectory(normalizePath(path).parent_path());
+    uncommitted_state.useDirectory(normalizePath(path).parent_path());
 
     operations.addOperation(std::make_unique<MetadataStorageFromPlainObjectStorageUnlinkMetadataFileOperation>(
         path,
@@ -532,8 +530,8 @@ void MetadataStorageFromPlainRewritableObjectStorageTransaction::removeRecursive
 
 void MetadataStorageFromPlainRewritableObjectStorageTransaction::createHardLink(const std::string & path_from, const std::string & path_to)
 {
-    uncommitted_state.lookupDirectory(normalizePath(path_from).parent_path());
-    uncommitted_state.lookupDirectory(normalizePath(path_to).parent_path());
+    uncommitted_state.useDirectory(normalizePath(path_from).parent_path());
+    uncommitted_state.useDirectory(normalizePath(path_to).parent_path());
 
     operations.addOperation(std::make_unique<MetadataStorageFromPlainObjectStorageCopyFileOperation>(
         path_from,
@@ -546,8 +544,8 @@ void MetadataStorageFromPlainRewritableObjectStorageTransaction::createHardLink(
 
 void MetadataStorageFromPlainRewritableObjectStorageTransaction::moveFile(const std::string & path_from, const std::string & path_to)
 {
-    uncommitted_state.lookupDirectory(normalizePath(path_from).parent_path());
-    uncommitted_state.lookupDirectory(normalizePath(path_to).parent_path());
+    uncommitted_state.useDirectory(normalizePath(path_from).parent_path());
+    uncommitted_state.useDirectory(normalizePath(path_to).parent_path());
 
     operations.addOperation(std::make_unique<MetadataStorageFromPlainObjectStorageMoveFileOperation>(
         /*replaceable=*/false,
@@ -562,8 +560,8 @@ void MetadataStorageFromPlainRewritableObjectStorageTransaction::moveFile(const 
 
 void MetadataStorageFromPlainRewritableObjectStorageTransaction::replaceFile(const std::string & path_from, const std::string & path_to)
 {
-    uncommitted_state.lookupDirectory(normalizePath(path_from).parent_path());
-    uncommitted_state.lookupDirectory(normalizePath(path_to).parent_path());
+    uncommitted_state.useDirectory(normalizePath(path_from).parent_path());
+    uncommitted_state.useDirectory(normalizePath(path_to).parent_path());
 
     operations.addOperation(std::make_unique<MetadataStorageFromPlainObjectStorageMoveFileOperation>(
         /*replaceable=*/true,
@@ -582,10 +580,22 @@ ObjectStorageKey MetadataStorageFromPlainRewritableObjectStorageTransaction::gen
     if (normalized_path.filename().empty())
         throw Exception(ErrorCodes::LOGICAL_ERROR, "File name is empty for path '{}'", path);
 
-    /// Materialize virtual parent.
     const auto parent_path = normalized_path.parent_path();
-    if (const auto [parent_exists, parent_info] = uncommitted_state.lookupDirectory(parent_path); parent_exists && !parent_info)
+    const auto [parent_exists, parent_info] = uncommitted_state.lookupDirectory(parent_path);
+
+    if (!parent_info)
+    {
+        /// Validate during commit that directory will be created on S3.
+        uncommitted_state.useMissingDirectory(parent_path);
+
+        /// Materialize virtual parent.
         createDirectoryRecursive(parent_path);
+    }
+    else
+    {
+        /// Validate during commit that directory will not be recreated on S3.
+        uncommitted_state.useDirectory(parent_path);
+    }
 
     if (const auto directory_remote_info = uncommitted_state.lookupDirectory(parent_path).second)
         return ObjectStorageKey::createAsAbsolute(metadata_storage.layout->constructFileObjectKey(directory_remote_info->remote_path, normalized_path.filename()));

--- a/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/MetadataStorageFromPlainRewritableObjectStorage.cpp
+++ b/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/MetadataStorageFromPlainRewritableObjectStorage.cpp
@@ -295,13 +295,13 @@ bool MetadataStorageFromPlainRewritableObjectStorage::existsFile(const std::stri
 
 bool MetadataStorageFromPlainRewritableObjectStorage::existsDirectory(const std::string & path) const
 {
-    return fs.takeReadOnlySnapshot()->existsDirectory(path).first;
+    return fs.takeReadOnlySnapshot()->existsDirectory(path);
 }
 
 bool MetadataStorageFromPlainRewritableObjectStorage::existsFileOrDirectory(const std::string & path) const
 {
     const auto tree = fs.takeReadOnlySnapshot();
-    return tree->existsFile(path) || tree->existsDirectory(path).first;
+    return tree->existsFile(path) || tree->existsDirectory(path);
 }
 
 uint64_t MetadataStorageFromPlainRewritableObjectStorage::getFileSize(const std::string & path) const
@@ -372,8 +372,10 @@ std::optional<Poco::Timestamp> MetadataStorageFromPlainRewritableObjectStorage::
 {
     const auto tree = fs.takeReadOnlySnapshot();
 
-    if (auto [exists, remote_info] = tree->existsDirectory(path); exists)
+    if (tree->existsDirectory(path))
     {
+        const auto remote_info = tree->getDirectoryRemoteInfo(path);
+
         if (remote_info)
             return Poco::Timestamp::fromEpochTime(remote_info->last_modified);
 
@@ -447,7 +449,7 @@ void MetadataStorageFromPlainRewritableObjectStorageTransaction::createDirectory
     operations.addOperation(std::make_unique<MetadataStorageFromPlainObjectStorageCreateDirectoryOperation>(
         /*recursive=*/false,
         normalizeDirectoryPath(path),
-        uncommitted_state.lookupDirectory(path).second->remote_path,
+        uncommitted_state.getDirectoryRemoteInfo(path)->remote_path,
         commit_snapshot,
         metadata_storage.object_storage,
         metadata_storage.layout,
@@ -467,7 +469,7 @@ void MetadataStorageFromPlainRewritableObjectStorageTransaction::createDirectory
     operations.addOperation(std::make_unique<MetadataStorageFromPlainObjectStorageCreateDirectoryOperation>(
         /*recursive=*/true,
         normalizeDirectoryPath(path),
-        uncommitted_state.lookupDirectory(path).second->remote_path,
+        uncommitted_state.getDirectoryRemoteInfo(path)->remote_path,
         commit_snapshot,
         metadata_storage.object_storage,
         metadata_storage.layout,
@@ -581,7 +583,7 @@ ObjectStorageKey MetadataStorageFromPlainRewritableObjectStorageTransaction::gen
         throw Exception(ErrorCodes::LOGICAL_ERROR, "File name is empty for path '{}'", path);
 
     const auto parent_path = normalized_path.parent_path();
-    const auto [parent_exists, parent_info] = uncommitted_state.lookupDirectory(parent_path);
+    const auto parent_info = uncommitted_state.getDirectoryRemoteInfo(parent_path);
 
     if (!parent_info)
     {
@@ -597,7 +599,7 @@ ObjectStorageKey MetadataStorageFromPlainRewritableObjectStorageTransaction::gen
         uncommitted_state.useDirectory(parent_path);
     }
 
-    if (const auto directory_remote_info = uncommitted_state.lookupDirectory(parent_path).second)
+    if (const auto directory_remote_info = uncommitted_state.getDirectoryRemoteInfo(parent_path))
         return ObjectStorageKey::createAsAbsolute(metadata_storage.layout->constructFileObjectKey(directory_remote_info->remote_path, normalized_path.filename()));
 
     throw Exception(ErrorCodes::LOGICAL_ERROR, "Directory '{}' does not exist", parent_path.string());

--- a/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/MetadataStorageFromPlainRewritableObjectStorageOperations.cpp
+++ b/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/MetadataStorageFromPlainRewritableObjectStorageOperations.cpp
@@ -80,20 +80,19 @@ MetadataStorageFromPlainObjectStorageCreateDirectoryOperation::MetadataStorageFr
 
 void MetadataStorageFromPlainObjectStorageCreateDirectoryOperation::execute()
 {
-    const auto [exists_directory, info] = fs_tree->existsDirectory(path);
-    if (info)
+    if (fs_tree->getDirectoryRemoteInfo(path))
         return;
 
     if (fs_tree->existsFile(path))
         throw Exception(ErrorCodes::CANNOT_CREATE_DIRECTORY, "File '{}' already exists", path.parent_path());
 
     if (!recursive)
-        if (!fs_tree->existsDirectory(path.parent_path().parent_path()).first)
+        if (!fs_tree->existsDirectory(path.parent_path().parent_path()))
             throw Exception(ErrorCodes::DIRECTORY_DOESNT_EXIST, "Directory '{}' does not exist", path.parent_path().parent_path());
 
     auto metadata_object_key = layout->constructDirectoryObjectKey(directory_remote_path);
 
-    if (exists_directory)
+    if (fs_tree->existsDirectory(path))
         LOG_TRACE(
             getLogger("MetadataStorageFromPlainObjectStorageCreateDirectoryOperation"),
             "Materializing virtual directory '{}' with remote path='{}'",
@@ -217,9 +216,9 @@ void MetadataStorageFromPlainObjectStorageMoveDirectoryOperation::execute()
     constexpr bool validate_content = false;
 #endif
 
-    if (!fs_tree->existsDirectory(path_from).first)
+    if (!fs_tree->existsDirectory(path_from))
         throw Exception(ErrorCodes::DIRECTORY_DOESNT_EXIST, "Directory '{}' does not exist", path_from);
-    else if (fs_tree->existsDirectory(path_to).first)
+    else if (fs_tree->existsDirectory(path_to))
         throw Exception(ErrorCodes::DIRECTORY_ALREADY_EXISTS, "Directory '{}' already exists", path_to);
     else if (normalizePath(path_from).empty())
         throw Exception(ErrorCodes::LOGICAL_ERROR, "Can't move root folder");
@@ -281,16 +280,14 @@ MetadataStorageFromPlainObjectStorageRemoveDirectoryOperation::MetadataStorageFr
 
 void MetadataStorageFromPlainObjectStorageRemoveDirectoryOperation::execute()
 {
-    auto [exists, remote_info] = fs_tree->existsDirectory(path);
-    if (!exists)
+    if (!fs_tree->existsDirectory(path))
         throw Exception(ErrorCodes::DIRECTORY_DOESNT_EXIST, "Directory '{}' does not exist", path);
     else if (auto children = fs_tree->listDirectory(path); !children.empty())
         throw Exception(ErrorCodes::CANNOT_RMDIR, "Directory '{}' is not empty. Children: [{}]", path, fmt::join(children, ", "));
     else if (normalizePath(path).empty())
         return;
 
-    chassert(remote_info.has_value());
-    info = std::move(remote_info.value());
+    info = fs_tree->getDirectoryRemoteInfo(path).value();
 
     LOG_TRACE(getLogger("MetadataStorageFromPlainObjectStorageRemoveDirectoryOperation"), "Removing directory '{}'", path);
 
@@ -438,9 +435,9 @@ void MetadataStorageFromPlainObjectStorageCopyFileOperation::execute()
 
     if (!fs_tree->existsFile(path_from))
         throw Exception(ErrorCodes::FILE_DOESNT_EXIST, "Metadata object for the source path '{}' does not exist", path_from);
-    else if (auto [exists, remote_info] = fs_tree->existsDirectory(path_to.parent_path()); !exists)
+    else if (!fs_tree->existsDirectory(path_to.parent_path()))
         throw Exception(ErrorCodes::DIRECTORY_DOESNT_EXIST, "Directory '{}' does not exist", path_to.parent_path());
-    else if (!remote_info.has_value())
+    else if (!fs_tree->getDirectoryRemoteInfo(path_to.parent_path()))
         throw Exception(ErrorCodes::LOGICAL_ERROR, "Directory '{}' is virtual", path_to.parent_path());
     else if (fs_tree->existsFile(path_to))
         throw Exception(ErrorCodes::FILE_ALREADY_EXISTS, "Target file '{}' already exists", path_to);
@@ -504,9 +501,9 @@ void MetadataStorageFromPlainObjectStorageMoveFileOperation::execute()
 
     if (!fs_tree->existsFile(path_from))
         throw Exception(ErrorCodes::FILE_DOESNT_EXIST, "File '{}' does not exist", path_from);
-    else if (auto [exists, remote_info] = fs_tree->existsDirectory(path_to.parent_path()); !exists)
+    else if (!fs_tree->existsDirectory(path_to.parent_path()))
         throw Exception(ErrorCodes::DIRECTORY_DOESNT_EXIST, "Directory '{}' does not exist", path_to.parent_path());
-    else if (!remote_info.has_value())
+    else if (!fs_tree->getDirectoryRemoteInfo(path_to.parent_path()))
         throw Exception(ErrorCodes::LOGICAL_ERROR, "Directory '{}' is virtual", path_to.parent_path());
 
     const auto normalized_path_from = normalizePath(path_from);
@@ -653,7 +650,7 @@ void MetadataStorageFromPlainObjectStorageRemoveRecursiveOperation::execute()
     if (normalizePath(path).empty())
         return;
 
-    if (fs_tree->existsDirectory(path).first)
+    if (fs_tree->existsDirectory(path))
     {
         move_tried = true;
         move_to_tmp_op->execute();

--- a/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/Transactions/UncommittedState.cpp
+++ b/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/Transactions/UncommittedState.cpp
@@ -85,23 +85,6 @@ private:
     std::unordered_set<std::string> created_directories;
 };
 
-void UncommittedState::useDirectory(const std::string & path) const
-{
-    const auto info = tx_snapshot->getDirectoryRemoteInfo(path);
-    if (!info)
-        return;
-
-    const auto snapshot_path = path_resolver->resolveToSnapshotPath(normalizePath(path));
-    if (snapshot_path && !path_resolver->isCreatedByTransaction(*snapshot_path))
-        preconditions->checkDirectoryPresent(*snapshot_path, info->remote_path);
-}
-
-void UncommittedState::useMissingDirectory(const std::string & path) const
-{
-    if (const auto snapshot_path = path_resolver->resolveToSnapshotPath(normalizePath(path)))
-        preconditions->checkDirectoryMissing(*snapshot_path);
-}
-
 UncommittedState::UncommittedState(std::shared_ptr<FsSnapshot> tx_snapshot_)
     : tx_snapshot(std::move(tx_snapshot_))
     , preconditions(std::make_shared<Preconditions>())
@@ -109,15 +92,38 @@ UncommittedState::UncommittedState(std::shared_ptr<FsSnapshot> tx_snapshot_)
 {
 }
 
+void UncommittedState::useDirectory(const std::string & path) const
+{
+    const auto info = tx_snapshot->getDirectoryRemoteInfo(path);
+    if (!info)
+        return;
+
+    const auto snapshot_path = path_resolver->resolveToSnapshotPath(normalizePath(path));
+    if (!snapshot_path)
+        return;
+
+    if (path_resolver->isCreatedByTransaction(*snapshot_path))
+        preconditions->checkDirectoryMissing(*snapshot_path);
+    else
+        preconditions->checkDirectoryPresent(*snapshot_path, info->remote_path);
+}
+
+void UncommittedState::useMissingDirectory(const std::string & path) const
+{
+    const auto info = tx_snapshot->getDirectoryRemoteInfo(path);
+    if (info)
+        return;
+
+    const auto snapshot_path = path_resolver->resolveToSnapshotPath(normalizePath(path));
+    if (snapshot_path)
+        preconditions->checkDirectoryMissing(*snapshot_path);
+}
+
 void UncommittedState::createDirectory(const std::string & path)
 {
     if (tx_snapshot->getDirectoryRemoteInfo(path))
-    {
-        useDirectory(path);
         return;
-    }
 
-    useMissingDirectory(path);
     path_resolver->recordCreate(normalizePath(path));
     tx_snapshot->recordDirectoryPath(path, DirectoryRemoteInfo{ .remote_path = getRandomASCIIString(32), .etag = "", .files = {}});
 }
@@ -125,12 +131,8 @@ void UncommittedState::createDirectory(const std::string & path)
 void UncommittedState::removeDirectory(const std::string & path)
 {
     if (!tx_snapshot->existsDirectory(path).first)
-    {
-        useMissingDirectory(path);
         return;
-    }
 
-    useDirectory(path);
     path_resolver->recordRemove(normalizePath(path));
     tx_snapshot->removeDirectory(path);
 }
@@ -138,25 +140,17 @@ void UncommittedState::removeDirectory(const std::string & path)
 void UncommittedState::moveDirectory(const std::string & path_from, const std::string & path_to)
 {
     if (!tx_snapshot->existsDirectory(path_from).first)
-    {
-        useMissingDirectory(path_from);
         return;
-    }
 
     if (tx_snapshot->existsDirectory(path_to).first || tx_snapshot->existsFile(path_to))
-    {
-        useDirectory(path_to);
         return;
-    }
 
-    useDirectory(path_from);
     path_resolver->recordMove(normalizePath(path_from), normalizePath(path_to));
     tx_snapshot->moveDirectory(path_from, path_to);
 }
 
 std::pair<bool, std::optional<DirectoryRemoteInfo>> UncommittedState::lookupDirectory(const std::string & path) const
 {
-    useDirectory(path);
     return tx_snapshot->existsDirectory(path);
 }
 

--- a/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/Transactions/UncommittedState.cpp
+++ b/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/Transactions/UncommittedState.cpp
@@ -130,7 +130,7 @@ void UncommittedState::createDirectory(const std::string & path)
 
 void UncommittedState::removeDirectory(const std::string & path)
 {
-    if (!tx_snapshot->existsDirectory(path).first)
+    if (!tx_snapshot->existsDirectory(path))
         return;
 
     path_resolver->recordRemove(normalizePath(path));
@@ -139,19 +139,19 @@ void UncommittedState::removeDirectory(const std::string & path)
 
 void UncommittedState::moveDirectory(const std::string & path_from, const std::string & path_to)
 {
-    if (!tx_snapshot->existsDirectory(path_from).first)
+    if (!tx_snapshot->existsDirectory(path_from))
         return;
 
-    if (tx_snapshot->existsDirectory(path_to).first || tx_snapshot->existsFile(path_to))
+    if (tx_snapshot->existsDirectory(path_to) || tx_snapshot->existsFile(path_to))
         return;
 
     path_resolver->recordMove(normalizePath(path_from), normalizePath(path_to));
     tx_snapshot->moveDirectory(path_from, path_to);
 }
 
-std::pair<bool, std::optional<DirectoryRemoteInfo>> UncommittedState::lookupDirectory(const std::string & path) const
+std::optional<DirectoryRemoteInfo> UncommittedState::getDirectoryRemoteInfo(const std::string & path) const
 {
-    return tx_snapshot->existsDirectory(path);
+    return tx_snapshot->getDirectoryRemoteInfo(path);
 }
 
 std::shared_ptr<Preconditions> UncommittedState::getTxPreconditions() const

--- a/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/Transactions/UncommittedState.h
+++ b/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/Transactions/UncommittedState.h
@@ -10,11 +10,11 @@ class UncommittedState
 {
     class PathResolver;
 
-    void useDirectory(const std::string & path) const;
-    void useMissingDirectory(const std::string & path) const;
-
 public:
     explicit UncommittedState(std::shared_ptr<FsSnapshot> tx_snapshot_);
+
+    void useDirectory(const std::string & path) const;
+    void useMissingDirectory(const std::string & path) const;
 
     void createDirectory(const std::string & path);
     void removeDirectory(const std::string & path);

--- a/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/Transactions/UncommittedState.h
+++ b/src/Disks/DiskObjectStorage/MetadataStorages/PlainRewritable/Transactions/UncommittedState.h
@@ -20,7 +20,7 @@ public:
     void removeDirectory(const std::string & path);
     void moveDirectory(const std::string & path_from, const std::string & path_to);
 
-    std::pair<bool, std::optional<DirectoryRemoteInfo>> lookupDirectory(const std::string & path) const;
+    std::optional<DirectoryRemoteInfo> getDirectoryRemoteInfo(const std::string & path) const;
     std::shared_ptr<Preconditions> getTxPreconditions() const;
 
 private:

--- a/src/Disks/tests/gtest_metadata_plain_rewritable_disk.cpp
+++ b/src/Disks/tests/gtest_metadata_plain_rewritable_disk.cpp
@@ -1969,103 +1969,6 @@ TEST_F(MetadataPlainRewritableDiskTest, CreateExistingDirectory)
     EXPECT_EQ(readObject(object_storage, metadata->getStorageObjects("A/B/C/file").front().remote_path), "1");
 }
 
-TEST_F(MetadataPlainRewritableDiskTest, ValidateDirectoryPreconditions)
-{
-    thread_local_rng.seed(42);
-
-    auto metadata = getMetadataStorage("ValidateDirectoryPreconditions");
-    auto object_storage = getObjectStorage("ValidateDirectoryPreconditions");
-
-    {
-        auto tx = metadata->createTransaction();
-        tx->createDirectoryRecursive("A/B/C");
-        tx->commit(DB::NoCommitOptions{});
-    }
-
-    /// Reused remote path: the directory is removed by a concurrent transaction before the commit.
-    {
-        auto tx = metadata->createTransaction();
-        tx->createDirectory("A/B/C");
-
-        {
-            auto concurrent_tx = metadata->createTransaction();
-            concurrent_tx->removeDirectory("A/B/C");
-            concurrent_tx->commit(DB::NoCommitOptions{});
-        }
-
-        EXPECT_ANY_THROW(tx->commit(DB::NoCommitOptions{}));
-        EXPECT_FALSE(metadata->existsDirectory("A/B/C"));
-    }
-
-    /// Reused remote path: the directory is removed and recreated by concurrent transactions,
-    /// so its remote path changed.
-    {
-        {
-            auto setup_tx = metadata->createTransaction();
-            setup_tx->createDirectoryRecursive("A/B/C");
-            setup_tx->commit(DB::NoCommitOptions{});
-        }
-
-        auto tx = metadata->createTransaction();
-        tx->createDirectory("A/B/C");
-
-        {
-            auto concurrent_tx = metadata->createTransaction();
-            concurrent_tx->removeDirectory("A/B/C");
-            concurrent_tx->commit(DB::NoCommitOptions{});
-        }
-        {
-            auto concurrent_tx = metadata->createTransaction();
-            concurrent_tx->createDirectoryRecursive("A/B/C");
-            concurrent_tx->commit(DB::NoCommitOptions{});
-        }
-
-        auto remote_prefix = generateObjectKeyPrefixForDirectoryPath(metadata, "A/B/C/");
-        EXPECT_ANY_THROW(tx->commit(DB::NoCommitOptions{}));
-        EXPECT_TRUE(metadata->existsDirectory("A/B/C"));
-        EXPECT_EQ(generateObjectKeyPrefixForDirectoryPath(metadata, "A/B/C/"), remote_prefix);
-    }
-
-    /// New remote path: the directory is created by a concurrent transaction before the commit.
-    {
-        auto tx = metadata->createTransaction();
-        tx->createDirectory("X");
-
-        {
-            auto concurrent_tx = metadata->createTransaction();
-            concurrent_tx->createDirectory("X");
-            concurrent_tx->commit(DB::NoCommitOptions{});
-        }
-
-        auto remote_prefix = generateObjectKeyPrefixForDirectoryPath(metadata, "X/");
-        EXPECT_ANY_THROW(tx->commit(DB::NoCommitOptions{}));
-        EXPECT_TRUE(metadata->existsDirectory("X"));
-        EXPECT_EQ(generateObjectKeyPrefixForDirectoryPath(metadata, "X/"), remote_prefix);
-    }
-
-    /// A file write into an existing directory is validated at commit time as well.
-    {
-        auto tx = metadata->createTransaction();
-        size_t file_size = writeObject(object_storage, tx->generateObjectKeyForPath("X/file").serialize(), "1");
-        tx->createMetadataFile("X/file", {StoredObject("X/file", "file", file_size)});
-
-        {
-            auto concurrent_tx = metadata->createTransaction();
-            concurrent_tx->removeDirectory("X");
-            concurrent_tx->commit(DB::NoCommitOptions{});
-        }
-
-        EXPECT_ANY_THROW(tx->commit(DB::NoCommitOptions{}));
-        EXPECT_FALSE(metadata->existsFile("X/file"));
-        EXPECT_FALSE(metadata->existsDirectory("X"));
-    }
-
-    EXPECT_EQ(listAllBlobs("ValidateDirectoryPreconditions"), sorted(std::vector<std::string>({
-        "./ValidateDirectoryPreconditions/__meta/wcageakzukwtfkvkwibqrfhzrrlubsbg/prefix.path", /// A/B/C
-        "./ValidateDirectoryPreconditions/huevlfwzdbhkvtedpymtjpafjjjfynpz/file"                /// X/file
-    })));
-}
-
 TEST_F(MetadataPlainRewritableDiskTest, RecreateDirectoryInSameTransaction)
 {
     auto metadata = getMetadataStorage("RecreateDirectoryInSameTransaction");
@@ -2213,8 +2116,8 @@ TEST_F(MetadataPlainRewritableDiskTest, SnapshotDecisionPreservedOnConcurrentCha
         tx2->commit(DB::NoCommitOptions{});
     }
 
-    EXPECT_ANY_THROW(tx1->commit(DB::NoCommitOptions{}));
-    EXPECT_TRUE(metadata->existsDirectory("X"));
+    tx1->commit(DB::NoCommitOptions{});
+    EXPECT_FALSE(metadata->existsDirectory("X"));
 }
 
 TEST_F(MetadataPlainRewritableDiskTest, ConcurrentCreateUnderUnlinkFile)
@@ -2257,8 +2160,6 @@ TEST_F(MetadataPlainRewritableDiskTest, ConcurrentRemoveOfMoveTarget)
         tx->commit(DB::NoCommitOptions{});
     }
 
-    /// A move whose target existed in the snapshot must keep failing
-    /// after the target is removed concurrently.
     auto tx1 = metadata->createTransaction();
     tx1->moveDirectory("A", "B");
 
@@ -2268,8 +2169,34 @@ TEST_F(MetadataPlainRewritableDiskTest, ConcurrentRemoveOfMoveTarget)
         tx2->commit(DB::NoCommitOptions{});
     }
 
-    EXPECT_ANY_THROW(tx1->commit(DB::NoCommitOptions{}));
+    tx1->commit(DB::NoCommitOptions{});
+
+    EXPECT_FALSE(metadata->existsDirectory("A"));
+    EXPECT_TRUE(metadata->existsDirectory("B"));
+}
+
+TEST_F(MetadataPlainRewritableDiskTest, ConcurrentCreateDirectory)
+{
+    auto metadata = getMetadataStorage("ConcurrentCreateDirectory");
+
+    /// Regression test for https://github.com/ClickHouse/ClickHouse/issues/111289
+    auto tx1 = metadata->createTransaction();
+    tx1->createDirectoryRecursive("A");
+
+    {
+        auto tx2 = metadata->createTransaction();
+        tx2->createDirectoryRecursive("A");
+        tx2->commit(DB::NoCommitOptions{});
+    }
+
+    auto remote_prefix = generateObjectKeyPrefixForDirectoryPath(metadata, "A/");
+    tx1->commit(DB::NoCommitOptions{});
 
     EXPECT_TRUE(metadata->existsDirectory("A"));
-    EXPECT_FALSE(metadata->existsDirectory("B"));
+    EXPECT_EQ(generateObjectKeyPrefixForDirectoryPath(metadata, "A/"), remote_prefix);
+    EXPECT_EQ(listAllBlobs("ConcurrentCreateDirectory").size(), 1);
+
+    metadata = restartMetadataStorage("ConcurrentCreateDirectory");
+    EXPECT_TRUE(metadata->existsDirectory("A"));
+    EXPECT_EQ(generateObjectKeyPrefixForDirectoryPath(metadata, "A/"), remote_prefix);
 }

